PYIC-8258: store reuse identities with PUT evcs endpoint

thebauSoftwire opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed
Store reuse identities (inc inherited identities)

### Why did it change
Returning users will have valid credentials that will need to have a corresponding Stored Identity Record. During phase 1, we aren't able to query SIS so IPV Core must assume that a returning user does not have an existing record and write one proactively so that when SIS is rolled out, returning users do not have to go through an identity proofing journey again.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-8258](https://govukverify.atlassian.net/browse/PYIC-8258)

## Checklists

<!-- Delete if changes don't include risk of credentials being exposed -->
- [x] No risk of PII, credentials or anything else sensitive being exposed through logs

<!-- Delete if changes don't apply -->
- [x] API/unit/contract tests have been written/updated


[PYIC-8258]: https://govukverify.atlassian.net/browse/PYIC-8258?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

thebauSoftwire committed:
PYIC-8258: store reuse identities and add api tests

thebauSoftwire committed:
PYIC-8258: self-review

thebauSoftwire committed:
PYIC-8258: fix failing api tests

thebauSoftwire committed:
Merge branch 'main' into pyic-8258

thebauSoftwire committed:
PYIC-8258: fix failing unit test

undermad commented on lambdas/process-candidate-identity/src/main/java/uk/gov/di/ipv/core/processcandidateidentity/ProcessCandidateIdentityHandler.java:
```diff
@@ -570,17 +563,25 @@ private void storeCandidateIdentity(
-                || STORE_IDENTITY_TYPES.contains(processIdentityType)
+                || shouldStoreIdentity(processIdentityType)
                 || PENDING.equals(processIdentityType);
     }
```
This become hard to read boolean logic to me. I would suggest something like this.
   
```
private boolean requiresVotMatchingResult(CandidateIdentityType processIdentityType) {
        if (configService.enabled(STORED_IDENTITY_SERVICE) && EXISTING.equals(processIdentityType)) {
            return true;
        }

        var combinedList = Stream.concat(PROFILE_MATCHING_TYPES.stream(), STORE_IDENTITY_TYPES.stream())
                .filter(identityType -> !identityType.equals(PENDING))
                .distinct()
                .toList();

        return combinedList.contains(processIdentityType);
    }

    private boolean requiresExistingVcsFromEvcs(CandidateIdentityType processIdentityType) {
        if (configService.enabled(STORED_IDENTITY_SERVICE) && EXISTING.equals(processIdentityType)) {
            return true;
        }

        return COI_CHECK_TYPES.contains(processIdentityType) || STORE_IDENTITY_TYPES.contains(processIdentityType);
    }
```

thebauSoftwire replied:
thanks for this :) have made similar changes but changed `configService.enabled(STORED_IDENTITY_SERVICE) && EXISTING.equals(processIdentityType)` to use the function below (`shouldStoreExistingIdentity`) as I thought it was more readable. With this, `requiresExistingVcsFromEvcs` could also be simplified, thanks!

undermad commented on api-tests/features/stored-identity/p2-journeys.feature:
```diff
@@ -113,3 +113,18 @@ Feature: Stored Identity - P2 journeys
+      | dcmaw   | kenneth-passport-valid |
+      | address | kenneth-current        |
+      | fraud   | kenneth-score-2        |
+
```
Can we add additional line that check the SIS record doesn't exist. This may be more verbose and indicate intention of that test.

undermad reviewed (changes requested):
Looking great, just few comments.

thebauSoftwire committed:
PYIC-8258: review markups:
- refactor conditional checks
- add step to check for no SIS record in SI reuse api test
